projects/graph/graph.py

```diff
@@ -7,42 +7,112 @@ class Graph:
     """Represent a graph as a dictionary of vertices mapping labels to edges."""
     def __init__(self):
         self.vertices = {}
+        self.queue = Queue()
+        self.stack = Stack()
     def add_vertex(self, vertex):
         """
         Add a vertex to the graph.
         """
-        pass  # TODO
+        self.vertices[vertex] = set()
     def add_edge(self, v1, v2):
         """
         Add a directed edge to the graph.
         """
-        pass  # TODO
+        if v1 not in self.vertices:
+            self.add_vertex(v1)
+        else:
+            self.vertices[v1].add(v2)
+
     def bft(self, starting_vertex):
         """
         Print each vertex in breadth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+
+        visited = []
+        vertices = self.vertices
+        queue = self.queue
+        queue.enqueue(starting_vertex)
+        while queue.size():
+            cur_ver = queue.dequeue()
+            if cur_ver not in visited:
+                visited.append(cur_ver)
+                for v in vertices[cur_ver]:
+                    queue.enqueue(v)
+        self.queue = Queue()
+        return visited
+
     def dft(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+        stack = self.stack
```

Comment:
You may be able to assign something like `stack = Stack()` on line 50. The stack can be independent of the graph instance. Vertices would require the usage of `self` because it's an attribute of the graph class. You can still `push`, `pop` and `iterate` over your Stack without using self, but your code is very explicit and easy to read 👍 Just something to tinker around with. __UPDATE__ I missed your Graph class `init()` code where you added the queue and stack classes. That's a great idea 🔥